scaladoc/src/dotty/tools/scaladoc/renderers/Renderer.scala

```diff
@@ -45,7 +45,52 @@ abstract class Renderer(rootPackage: Member, val members: Map[DRI, Member], prot
           val msg = s"ERROR: Multiple index pages for doc found ${indexes.map(_.file)}"
           report.error(msg)
 
-        val templatePages = templates.map(templateToPage(_, siteContext))
+        // Below code is for walking in order the tree and modifing its nodes basing on its neighbours
+
+        // First we flatten templates to get them sorted in-order
+        def flattenedTemplates(template: LoadedTemplate): Seq[LoadedTemplate] =
+          template +: template.children.flatMap(flattenedTemplates)
+
+        // We add dummy guards
+        val allTemplates: Seq[Option[LoadedTemplate]] = None +: templates.flatMap(flattenedTemplates).map(Some(_)) :+ None
+
+        // Let's gather the list of maps for each template with its in-order neighbours
+        val newSettings: Seq[Map[String, Object]] = allTemplates.sliding(size = 3, step = 1).map {
+          case None :: None :: Nil =>
+            Map.empty
+          case prev :: mid :: next :: Nil =>
+            def link(sibling: Option[LoadedTemplate]): Option[String] =
+              def realPath(path: Path) = if Files.isDirectory(path) then Paths.get(path.toString, "index.html") else path
+              sibling.map { n =>
+                val realMidPath = realPath(mid.get.file.toPath)
+                val realSiblingPath = realPath(n.file.toPath)
+                realMidPath.relativize(realSiblingPath).toString.stripPrefix("../")
+              }
+            List(link(prev).map("previous" -> _), link(next).map("next" -> _)).flatten.toMap
+        }.toSeq
+
+        // We update the immutable tree of templates by walking in-order
+        def updateSettings(template: LoadedTemplate, additionalSettings: Seq[Map[String, Object]]): (LoadedTemplate, Seq[Map[String, Object]]) =
```

Comment:
I have hard time understanding the implementation of this method. I think it needs thorough tests, or maybe a rewrite. I wonder why we have to introduce a `null` value.

What about something like the following?

~~~ scala
def updateSettings(templates: Seq[LoadedTemplate], additionalSettings: List[Map[String, Object]]): (Seq[LoadedTemplate], List[Map[String, Object]]) =
    val updatedTemplates = List.newBuilder[LoadedTemplate]
    var remainingSettings = additionalSettings
    for (template <- templates) do
      val updatedTemplateFile = template.templateFile.settings.updated("page", remainingSettings.head ++ template.templateFile.settings.getOrElse("page", Map.empty))
      remainingSettings = remainingSettings.tail
      updatedTemplates += template.copy(
        templateFile = updatedTemplateFile,
        children = updateSettings(template.children, remainingSettings)
      )
    (updatedTemplates.result(), remainingSettings)
~~~

Not sure this is better either :)
I wonder if it would be better to do it in a single pass instead of first computing `newSettings` and then performing `updateSettings`.

Reply:
I know it's not super clear, I also had a hard time designing it, but I tried to stick to immutable collections, maybe wrongly. The null value is for `scanLeft` since it needs initial value, though in our lambda we don't use it (`case ((_, as), template)`). Will reconsider changing it, maybe I would simplify it even more if I will use mutable state.

Reply:
I was thinking how to do that in a single run, but the problem is very hard without flattening the pages, because we would have to calculate the `next` and `previous` page, which is quite hard. In theory, finding predecessor and successor is O(n), however, we would have to zip it with indexes or implement bidirectional iterators, pass explicitly the parent etc. U added your snippet of code (changed it a little), and I think it's the best what we can do in terms of readability and efficiency